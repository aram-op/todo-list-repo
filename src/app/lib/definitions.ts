export type Todo =  {
    id: string,
    title: string,
    is_completed: boolean
}