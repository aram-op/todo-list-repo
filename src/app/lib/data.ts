import {createPool} from '@vercel/postgres';
import {Todo} from '@/app/lib/definitions';

const pool = createPool({
    connectionString: process.env.NEXT_PUBLIC_POSTGRES_URL,
});

export async function fetchFilteredTodos(query: string) {

    try {
        const todos = await pool.sql<Todo>`
            SELECT * FROM todos WHERE UPPER(title) LIKE UPPER(${`%${query}%`});
        `;

        return todos.rows;
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todos.');
    }
}

export async function fetchTodoById(id: string) {
    try {
        const result = await pool.sql<Todo>`
            SELECT * FROM todos WHERE id = ${id};
        `;

        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todo.');
    }
}

export async function updateTodo(todo: Todo) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${todo.title}`},
             is_completed = ${`${todo.is_completed}`}
             WHERE id = ${`${todo.id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function updateTodoTitle(id: string, title: string) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${title}`},
             WHERE id = ${`${id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function deleteTodoById(id: string) {
    try {
        const result = await pool.sql`
              DELETE FROM todos WHERE id = ${`${id}`};
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to delete todo.');
    }
}

export async function addTodo(title: string) {
    try {
        const result = await pool.sql`
            INSERT INTO todos (title, is_completed)
             VALUES
             (${title}, false)
             ON CONFLICT (id) DO NOTHING;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to add todo.')
    }
}