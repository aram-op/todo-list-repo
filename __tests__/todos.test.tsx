import { render, screen, waitFor } from '@testing-library/react';
import { fetchFilteredTodos } from '@/app/lib/data';
import { Todo } from '@/app/lib/definitions';
import Todos from '@/app/ui/todos/todos';

jest.mock('@/app/lib/data', () => ({
    fetchFilteredTodos: jest.fn(),
}));

jest.mock('@/app/ui/todo-list-item/todo-list-item', () => {
    return jest.fn(({ todo, onItemRemoved }) => (
        <li data-testid="todo-item">
            <p>{todo.title}</p>
            <button onClick={() => onItemRemoved(todo.id)}>Remove</button>
        </li>
    ));
});

describe('Todos component', () => {
    const mockData: Todo[] = [
        { id: '1', title: 'Buy milk', is_completed: false },
        { id: '2', title: 'Go for a walk', is_completed: true },
    ];

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('it should fetch and display todos', async () => {
        (fetchFilteredTodos as jest.Mock).mockResolvedValue(mockData);

        render(<Todos query="test-query" />);

        await waitFor(() => expect(screen.getAllByTestId('todo-item')).toHaveLength(2));

        expect(screen.getByText('Buy milk')).toBeInTheDocument();
        expect(screen.getByText('Go for a walk')).toBeInTheDocument();
        expect(fetchFilteredTodos).toHaveBeenCalledWith('test-query');
    });

    it('should handle todo removing', async () => {
        (fetchFilteredTodos as jest.Mock).mockResolvedValue(mockData);

        render(<Todos query="test-query" />);

        await waitFor(() => expect(screen.getAllByTestId('todo-item')).toHaveLength(2));

        const removeButton = screen.getAllByText('Remove')[0];
        removeButton.click();

        await waitFor(() => expect(screen.getAllByTestId('todo-item')).toHaveLength(1));
        expect(screen.queryByText('Buy milk')).not.toBeInTheDocument();
    });

    it('should match snapshot', async () => {
        (fetchFilteredTodos as jest.Mock).mockResolvedValue(mockData);

        const { asFragment } = render(<Todos query="test-query" />);
        await waitFor(() => expect(screen.getAllByTestId('todo-item')).toHaveLength(2));

        expect(asFragment()).toMatchSnapshot();
    });
});
