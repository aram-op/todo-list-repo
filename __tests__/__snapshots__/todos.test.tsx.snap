// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos component should match snapshot 1`] = `
<DocumentFragment>
  <ul
    class="list"
  >
    <li
      data-testid="todo-item"
    >
      <p>
        Buy milk
      </p>
      <button>
        Remove
      </button>
    </li>
    <li
      data-testid="todo-item"
    >
      <p>
        Go for a walk
      </p>
      <button>
        Remove
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
